// Copyright Â© Her Majesty the Queen in Right of Canada, as represented
// by the Minister of Statistics Canada, 2019.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
/* Protocol buffer implementation of physt classed.

Idea and first implementation by Ryan Mackenzie White
(ttps://github.com/ryanmackenziewhite).

Note: Work in progress and not all properties
    can be serialized. Future versions of the protocol
    will try to keep consistency. 


*/
syntax = "proto3";
package physt;

message Meta {
    string name = 1;
    string title = 2;
    repeated string axis_names = 3;
    /* map<string, string> user_defined = 4; */  // not supported yet
}

message Bin {
    double lower = 1;
    double upper = 2;
}

message Binning {
    string binning_type = 1;        // always 'StaticBinning' 
    repeated Bin bins = 2;
    /* bool adaptive = 1; */        // not supported yet
}

message Histogram {
    string physt_version = 1;
    string physt_compatible = 2;
    string histogram_type = 3;
    repeated Binning binnings = 4;    
    string dtype = 5;
    repeated double frequencies = 6;
    repeated double errors2 = 7;
    /* repeated double missed = 8;  // not supported yet
    bool missed_keep = 9; */        // not supported yet
    Meta meta = 10;
    
}

message HistogramCollection {
    map<string, Histogram> histograms = 1;
}

